### ISSUE TITLE
`global_asm!` in statement position has weird behavior with `Self`

### ISSUE THREAD

**Issue by asquared31415** [rust-lang/rust#156894]
```rust
use std::marker::PhantomData;
use std::arch::global_asm;

struct S<T>(PhantomData<T>);

impl<T> S<T> {
    fn meow(&self) {
        global_asm!("/* {} */", sym Self::meow);
    }
}
```

This code should be forbidden because it uses a generic in `global_asm!`, which isn't meaningful.

Instead, the compiler became confused and was reporting errors about `Sized`.

```
error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> src/lib.rs:8:37
  |
6 | impl<T> S<T> {
  |      - this type parameter needs to be `Sized`
7 |     fn meow(&self) {
8 |         global_asm!("/* {} */", sym Self::meow);
  |                                     ^^^^ doesn't have a size known at compile-time
  |
note: required by an implicit `Sized` bound in `S`
 --> src/lib.rs:4:10
  |
4 | struct S<T>(PhantomData<T>);
  |          ^ required by the implicit `Sized` requirement on this type parameter in `S`
help: consider relaxing the implicit `Sized` restriction
  |
4 | struct S<T: ?Sized>(PhantomData<T>);
  |           ++++++++

error[E0599]: the associated function or constant `meow` exists for struct `S<T>`, but its trait bounds were not satisfied
 --> src/lib.rs:8:43
  |
4 | struct S<T>(PhantomData<T>);
  | ----------- associated function or constant `meow` not found for this struct
...
8 |         global_asm!("/* {} */", sym Self::meow);
  |                                           ^^^^ associated function or constant cannot be called on `S<T>` due to unsatisfied trait bounds
  |
note: trait bound `T: Sized` was not satisfied
 --> src/lib.rs:6:6
  |
6 | impl<T> S<T> {
  |      ^  ----
  |      |
  |      unsatisfied trait bound introduced here
help: consider relaxing the type parameter's implicit `Sized` bound
  |
6 | impl<T: ?Sized> S<T> {
  |       ++++++++
```

### Meta
`rustc --version --verbose`:
```
rustc 1.98.0-nightly (23a3312d9 2026-05-23)
binary: rustc
commit-hash: 23a3312d92a1c4ba0373f1e25277be20ba8bb28c
commit-date: 2026-05-23
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.6
```

Caused by #156582 allowing this position
Potentially related to #156760, a variant that uses turbofish syntax instead
cc #156855 and #156884

@rustbot label +A-global-asm +T-compiler

**Comment by asquared31415** [comment]
It is good that this code is rejected, but I'm filing this because I believe it's being rejected for incoherent reasons and that there's underlying bugs instead.

**Comment by folkertdev** [comment]
cc @daxpedda another useful test case when you re-land the PR

**Comment by theemathas** [comment]
This code causes an ICE with `rustc 1.98.0-nightly (23a3312d9 2026-05-23)` (fixed in the current nightly, presumably due to #156884):

```rust
#![feature(sized_hierarchy)]

use std::arch::global_asm;
use std::marker::{PhantomData, PointeeSized};

struct S<T: PointeeSized>(PhantomData<T>);

impl<T: PointeeSized> S<T> {
    fn meow(&self) {
        global_asm!("/* {} */", sym Self::meow);
    }
}
```

<details>
<summary>Error</summary>

```
error: internal compiler error: /rustc-dev/23a3312d92a1c4ba0373f1e25277be20ba8bb28c/compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs:455:13: debuginfo: Trying to create type name for unexpected type: T/#0


thread 'rustc' (787529) panicked at /rustc-dev/23a3312d92a1c4ba0373f1e25277be20ba8bb28c/compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs:455:13:
Box<dyn Any>
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   2: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: rustc_codegen_ssa::debuginfo::type_names::push_debuginfo_type_name
   7: rustc_codegen_ssa::debuginfo::type_names::push_generic_args_internal
   8: rustc_codegen_ssa::debuginfo::type_names::push_debuginfo_type_name
   9: rustc_codegen_ssa::debuginfo::type_names::compute_debuginfo_type_name
  10: rustc_codegen_llvm::debuginfo::metadata::build_struct_type_di_node
  11: rustc_codegen_llvm::debuginfo::metadata::spanned_type_di_node
  12: rustc_codegen_llvm::debuginfo::metadata::build_pointer_or_reference_di_node
  13: rustc_codegen_llvm::debuginfo::metadata::spanned_type_di_node
  14: <rustc_codegen_llvm::context::GenericCx<rustc_codegen_llvm::context::FullCx> as rustc_codegen_ssa::traits::debuginfo::DebugInfoCodegenMethods>::dbg_scope_fn
  15: <rustc_codegen_llvm::context::GenericCx<rustc_codegen_llvm::context::FullCx> as rustc_codegen_ssa::traits::debuginfo::DebugInfoCodegenMethods>::create_function_debug_context
  16: rustc_codegen_ssa::mir::codegen_mir::<rustc_codegen_llvm::builder::GenericBuilder<rustc_codegen_llvm::context::FullCx>>
  17: rustc_codegen_llvm::base::compile_codegen_unit::module_codegen
  18: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::ExtraBackendMethods>::compile_codegen_unit
  19: rustc_codegen_ssa::base::codegen_crate::<rustc_codegen_llvm::LlvmCodegenBackend>
  20: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  21: <rustc_interface::queries::Linker>::codegen_and_build_linker
  22: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  23: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/Users/timch/foo/rustc-ice-2026-05-25T07_04_39-10752.txt` to your bug report

note: rustc 1.98.0-nightly (23a3312d9 2026-05-23) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `foo` (lib)
```

</details>
